server/src/internalClusterTest/java/org/elasticsearch/health/GetHealthCancellationIT.java

```diff
@@ -0,0 +1,157 @@
+
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST, numDataNodes = 0, numClientNodes = 0)
+public class GetHealthCancellationIT extends ESIntegTestCase {
+
+    @Override
+    protected Collection<Class<? extends Plugin>> nodePlugins() {
+        return Arrays.asList(getTestTransportPlugin(), MockTransportService.TestPlugin.class);
+    }
+
+    @Override
+    protected Settings nodeSettings(int ordinal, Settings otherSettings) {
+        return Settings.builder()
+            .put(super.nodeSettings(ordinal, otherSettings))
+            .put(NetworkModule.HTTP_DEFAULT_TYPE_SETTING.getKey(), Netty4Plugin.NETTY_HTTP_TRANSPORT_NAME)
+            .build();
+    }
+
+    @Override
+    protected boolean addMockHttpTransport() {
+        return false; // enable http
+    }
+
+    public void testCancellation() throws Exception {
+        internalCluster().startMasterOnlyNode(Settings.builder().put(Node.NODE_NAME_SETTING.getKey(), "master_node").build());
+        internalCluster().startDataOnlyNode(Settings.builder().put(Node.NODE_NAME_SETTING.getKey(), "data_node").build());
+
+        final CountDownLatch tasksBlockedLatch = new CountDownLatch(1);
+        final SubscribableListener<Void> fetchHealthInfoRequestReleaseListener = new SubscribableListener<>();
+        for (TransportService transportService : internalCluster().getInstances(TransportService.class)) {
+            ((MockTransportService) transportService).addRequestHandlingBehavior(
+                FetchHealthInfoCacheAction.NAME,
+                (handler, request, channel, task) -> {
+                    tasksBlockedLatch.countDown();
+                    fetchHealthInfoRequestReleaseListener.addListener(
+                        ActionListener.wrap(ignored -> handler.messageReceived(request, channel, task), e -> {
+                            throw new AssertionError("unexpected", e);
+                        })
+                    );
+                }
+            );
+        }
+
+        final ClusterService clusterService = internalCluster().getCurrentMasterNodeInstance(ClusterService.class);
+        final CountDownLatch healthNodeFoundLatch = new CountDownLatch(1);
+        AtomicReference<DiscoveryNode> healthNodeReference = new AtomicReference<>();
+        // the health node might take a bit of time to be assigned by the persistent task framework so we wait until we have a health
+        // node in the cluster before proceeding with the test
+        // proceeding with the execution before the health node assignment would yield a non-deterministic behaviour as we
+        // wouldn't call the transport service anymore (there wouldn't be a node to fetch the health information from)
+        final ClusterStateListener clusterStateListener = event -> {
+            setHealthNodeReferenceIfPresent(event.state(), healthNodeReference, healthNodeFoundLatch);
+        };
+        clusterService.addListener(clusterStateListener);
```

Comment:
hygiene nit: we should remove one-shot listeners when no longer needed. Doesn't really matter for now since we're in a dedicated cluster, but it can cause problems if the situation changes.